src/server/Commands.hs

```diff
@@ -260,6 +263,62 @@ cmdRetr sock state params
   where
     getSock (DataSocket sock) = sock
 
+cmdCdup :: Socket -> UserState -> [String] -> IO UserState
+cmdCdup sock state params
+  | getIsLoggedin state = do
+    if checkParams params 0
+      then do
+        rootDir <- makeAbsolute (getRootDir state)
+        currDir <- getCurrentDirectory 
+        if isSubdirectory rootDir currDir
+          then do
+            let parentDir = getParentDirectory currDir
+            setCurrentDirectory parentDir
+            sendLine sock ("200 Command okay. Current Directory: " ++ parentDir)
+            return state
+          else do
+            sendLine sock "550 Insufficient permissions. Action not taken."
+            return state
+      else do 
+        sendLine sock "501 Syntax error in parameters or arguments."
+        return state
+  | otherwise = do
+    sendLine sock "530 Not logged in."
+    return state
+
+cmdCwd :: Socket -> UserState -> [String] -> IO UserState
+cmdCwd sock state params
+  | getIsLoggedin state = do
+    if checkParams params 1
+      then do
+        if isRelative userPath
+          then do
+            absPath <- makeAbsolute userPath
+            cwdHelper sock state absPath
+          else do
+            cwdHelper sock state userPath
+      else do
+        sendLine sock "501 Syntax error in parameters or arguments."
+        return state
+  | otherwise = do
+    sendLine sock "530 Not logged in."
+    return state
+  where
+    userPath = getFirst params
+
+cwdHelper :: Socket -> UserState -> FilePath -> IO UserState
+cwdHelper sock state path = do
+  rootDir <- makeAbsolute (getRootDir state)
+  validDir <- doesDirectoryExist path
+  if validDir && (equalFilePath rootDir path || isSubdirectory rootDir path)
```

Comment:
`isSubdirectory rootDir path`
This will return true if the path is `../`
if currDir == rootDir = /home/cs/FTPServer
and I send the server:
`cwd ../`
It won't fail because the absolute path becomes
`/home/cs/FTPServer/..` which is equivalent to `home/cs/`

The easiest way to solve this is to reject `cwd` commands that have a path that starts with `../`

Reply:
Sending a relative path like `../` will work with the current code because it is made absolute before the cwdHelper function however I should also be making the path absolute regardless because an input like `/home/cs/FTPServer/../../` won't fail.